### PR TITLE
cpu/stm32_common: Bug fixes for stm i2c

### LINKED ISSUE #3366: periph/i2c: handle NACK
The I2C peripheral driver interface does not currently handle a slave `NACK`.

As there are I2C devices which have `NACK` as a valid part of the communication process, the current interface (which only returns `-1` on `undefined device given`) can not cope with these devices.

I'd suggest adding a `-2` for `no ACK received` to the interface documentation and changing the driver implementations accordingly.

### LINKED ISSUE #10395: stm32f103xx: I2C has no internal pullups and af config
#### Description
When using the nucleo-f103rb the standard configuration is i2c0 on PB8 and PB9 the AFIO periph clock must be enabled 
`RCC->APB2ENR |= 0x00000001;` 
and the pin must be remapped to those pins 
`AFIO->MAPR |= AFIO_MAPR_I2C1_REMAP;`
Also due to the hardware constraints it seems like the internal pullup cannot be enabled since the pins must be set to `GPIO_AF_OUT_OD`.

In the [reference manual](https://www.st.com/content/ccc/resource/technical/document/reference_manual/59/b9/ba/7f/11/af/43/d5/CD00171190.pdf/files/CD00171190.pdf/jcr:content/translations/en.CD00171190.pdf) it seems like it is impossible to do so (in section 9.1.9 Alternate function configuration it states that 'The weak pull-up and pull-down resistors are disabled.')

I am not sure if anything can be done about requiring external pullups.

#### Steps to reproduce the issue
Try to use the i2c0 on any nucleo103xx on PB8 and PB9.

#### Expected results
It works.

#### Actual results
It doesn't work and will hang

### PR DESCRIPTION
**Pull request by MrKevinWeiss**
### Contribution description
This PR fixes many bugs discovered by the HiL.
* Fixes remappable sda and scl pins on stm32f1xx boards
* Refactors and provides accurate features for both i2c_1 and i2c_2
* Uses common i2c functions when possible
* Documents hardware constraints
* Prevents or recovers from lockup states
* Reduces code size
* More accurately matches i2c API

_Example of bytes saved with build sizes of tests/periph_i2c_

BOARD | with PR | master (at the time) | bytes saved
-|-|-|-
nucleo-l073rz DEVELHELP=1 | 16004 | 16188 | 184
nucleo-l073rz DEVELHELP=0 | 14952 | 15140 | 188
nucleo-f410rb DEVELHELP=1 | 15936 | 15992 | 56
nucleo-f410rb DEVELHELP=0 | 14976 | 15000 | 24

### Testing procedure
Test on stm32f1, stm32f4 and (stm32f3, f0, l0 or l4)
One can run the robot-test from the HiL branch or run term in tests/periph_i2c/
Or run manually against your favorite i2c device.  No everything will be supported but everything should at least give correct error messages and not lock up the bus.
 *If nucleo-f103rb is used it needs external pullups.
1. Basic read regs with 1,2,3 bytes
2. Basic write regs with 1,2,3 bytes
3. Test split frame (NOSTOP, NOSTOP|NOSTART, NOSTART) reads 
4. Test split frame writes 
5. Repeated start write (write_bytes+NOSTOP, write_bytes)
6. Getting an address nack then read a proper register.

### Issues/PRs references
fixes #10395 
helps with stm32 based boards for #3366
checks some boxes in #9518

### CODE REVIEW

**Comment by MrKevinWeiss** [comment]
Still a WIP but I tried to really clean up the i2c_2.c since most things were not working.  I think I may have to also do the same for i2c_1.c, at least get it into a state that it can use the standard read/write regs.

I can try to setup so the HiL tests it too.

**Comment by MrKevinWeiss** [comment]
@aabadie ready for review, it is easier if you go by each commit.  They should all compile and be testable separately as suggested by @cladmi

File: cpu/stm32_common/periph/i2c_2.c [hunk 1]
```diff
@@ -16,9 +17,9 @@
  * @file
  * @brief       Low-level I2C driver implementation
  *
- * This driver supports the STM32 F4 families.
+ * This driver supports the STM32 F1, L1, and F4 families.
```
Comment (aabadie) [inline cpu/stm32_common/periph/i2c_2.c]: and F2

**Comment by aabadie** [comment]
Also it's quite unclear to me why this PR would fix #3366 since this issue is not only related to stm32. Can we confirm that it's already fixed for other families (nrf, etc) ? If it's not the case, I suggest to remove the mention in the initial comment if this PR.

**Comment by MrKevinWeiss** [comment]
@aabadie True, It only helps the stm, I have not check all the other boards.

File: cpu/stm32_common/periph/i2c_2.c (hunk 1 above)
Comment (MrKevinWeiss) [reply cpu/stm32_common/periph/i2c_2.c]: Yup thanks,  I will split everything out.

**Comment by MrKevinWeiss** [comment]
Closing in favor of #10608 #10610 #10607 
The pm layer was left out as it was only added to make things look the same and was not initially tested.  This may be added at a later date.

**Pull request closed without merging**